warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/FieldHarvester.java

```diff
@@ -0,0 +1,271 @@
+     * @param value
+     *            the record from which the fields will be extracted
+     * @param offset
+     *            record offset within the source file
+     * @param splitStart
+     *            the splitStart for the record
+     */
+    public void extractFields(Multimap<String,NormalizedContentInterface> fields, IngestHelperInterface ingestHelper, RawRecordContainer value, long offset,
+                    String splitStart) throws Exception {
+        // reset exception-in-extraction tracking
+        this.originalException = null;
+        
+        // "candidateFields" holds the fields that will eventually be added to "fields"
+        Multimap<String,NormalizedContentInterface> candidateFields;
+        
+        try {
+            // parse the record into its candidate field names and values using the IngestHelperInterface.
+            candidateFields = ingestHelper.getEventFields(value);
+        } catch (Exception exception) {
+            // delay throwing the exception to attempt salvaging
+            this.originalException = exception;
+            candidateFields = attemptToSalvageFields(value, ingestHelper);
+        }
+        
+        try {
+            // try adding supplemental fields to candidateFields, whether or not they were salvaged
+            addSupplementalFields(value, offset, splitStart, ingestHelper, candidateFields);
+        } catch (Exception exception) {
+            if (null == this.originalException) {
+                this.originalException = exception;
+            } else {
+                // preserve original exception and log the latest exception
+                log.error(exception);
+            }
+        }
+        
+        // add candidateFields to fields, even if there was an error
+        // identify if any individual fields contain an error
+        addFieldsAndDetectFieldErrors(fields, candidateFields);
+        
+        if (null != this.originalException) {
+            log.error("Rethrowing original exception after completing field extraction.");
+            throw originalException;
+        }
+    }
+    
+    @VisibleForTesting
+    boolean hasError() {
+        return null != this.originalException;
+    }
+    
+    @VisibleForTesting
+    Exception getOriginalException() {
+        return this.originalException;
+    }
+    
+    /**
+     * If IngestHelper implements FieldSalvager, get the salvageable fields from value. Otherwise, return an empty Multimap.
+     */
+    private Multimap<String,NormalizedContentInterface> attemptToSalvageFields(RawRecordContainer value, IngestHelperInterface ingestHelper) {
+        // If this helper is able, attempt to salvage a subset of the fields
+        if (null != ingestHelper && ingestHelper instanceof FieldSalvager) {
+            FieldSalvager salvager = (FieldSalvager) ingestHelper;
+            try {
+                Multimap<String,NormalizedContentInterface> salvagedFields = salvager.getSalvageableEventFields(value);
```

Comment:
I am wondering whether a better approach may be to pass the multimap into the IngestHelper.getFields(value) method such that if an exception is thrown, then the map will already contain the fields that could be "salvaged".  That way we are forced to allow allow ingest helpers to "salvage fields" and not have to implement a separate interface.